### PR TITLE
Real-time maps with EventHubs, SignalR, and LeafletJS

### PR DESCRIPTION
**Pull request by noodlefrenzy**
Case-study for http://www.mikelanzetta.com/2015/08/real-time-mapping-with-signalr-and-event-hubs/ and https://github.com/noodlefrenzy/asp-mappy

### CODE REVIEW

File: _posts/2015-08-26-Real-Time-Maps-with-Event-Hubs-and-SignalR.md [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+If we could measure engagement, we could work with the city of Hamburg to try and make the people happier, and measure results (this was my first inkling I wasn't working with a US company).
+
+These were grandiose notions, to be sure, but the first step was to visualize the data. 
+As with all geographic data, the only way to get a proper sense of what it looks like is to visualize it on a map, so we needed a way to get real-time data from mobile applications to our map visualization.
+We built a [Cordova](http://cordova.apache.org/)-based mobile app driving geo-data into [Event Hubs](http://azure.microsoft.com/en-us/services/event-hubs/) (for future data), and a C#-based CLI for driving their existing data into Event Hubs as a spoof "real-time" feed. 
+Now that we had our geo-data, we built an [EventProcessorHost](https://www.nuget.org/packages/Microsoft.Azure.ServiceBus.EventProcessorHost) to consume it from within our ASP.Net website, do any required pre-processing, and drive the resulting data through to listening clients via [SignalR](http://signalr.net/).
```
Comment (felixrieseberg) [inline _posts/2015-08-26-Real-Time-Maps-with-Event-Hubs-and-SignalR.md]: This is nitpicking to the extreme, but you *could* replace `geo-data` with `geodata`.

**Comment by felixrieseberg** [comment]
Good stuff! I'd appreciate it if you could squash your commits (`git rebase -i HEAD~6`, mark all the ones except for the first one as `fixup`, then update the PR with a `git push -f`), but otherwise: LGTM!

**Comment by sedouard** [comment]
FYI - Peer reviewing this one

File: _posts/2015-08-26-Real-Time-Maps-with-Event-Hubs-and-SignalR.md [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+Our main client was a single-page web application using [Leaflet.js](http://leafletjs.com/) to render the most recent paths, and per-person and aggregated heatmaps.
+
+The ASP.Net application was [released on GitHub](https://github.com/noodlefrenzy/asp-mappy), and a more detailed walkthrough can be found [on my blog](http://www.mikelanzetta.com/2015/08/real-time-mapping-with-signalr-and-event-hubs/).
+
+![Screenshot of ASP.Net application]({{ site.baseurl }}/images/2015-08-26-Real-Time-Maps-with-Event-Hubs-and-SignalR/route_animation.gif)
```
Comment (sedouard) [inline _posts/2015-08-26-Real-Time-Maps-with-Event-Hubs-and-SignalR.md]: broken link

File: _posts/2015-08-26-Real-Time-Maps-with-Event-Hubs-and-SignalR.md [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+Naturally, as in any mature industry, they had groups and standards providing rough values for these numbers, but real data was harder to come by and until now had required large surveys and estimation.
+If we could gather this data in real-time, we could not only more effectively target content to where it will be seen, but we could measure engagement. 
+If we could measure engagement, we could work with the city of Hamburg to try and make the people happier, and measure results (this was my first inkling I wasn't working with a US company).
+
+These were grandiose notions, to be sure, but the first step was to visualize the data. 
+As with all geographic data, the only way to get a proper sense of what it looks like is to visualize it on a map, so we needed a way to get real-time data from mobile applications to our map visualization.
+We built a [Cordova](http://cordova.apache.org/)-based mobile app driving geo-data into [Event Hubs](http://azure.microsoft.com/en-us/services/event-hubs/) (for future data), and a C#-based CLI for driving their existing data into Event Hubs as a spoof "real-time" feed.
```
Comment (sedouard) [inline _posts/2015-08-26-Real-Time-Maps-with-Event-Hubs-and-SignalR.md]: screenshots of the app would be nice here

**Comment by sedouard** [comment]
section headers would be nice to have to make things more visually appealing

File: _posts/2015-08-26-Real-Time-Maps-with-Event-Hubs-and-SignalR.md (hunk 2 above)
Comment (sedouard) [inline _posts/2015-08-26-Real-Time-Maps-with-Event-Hubs-and-SignalR.md]: broken link

File: _posts/2015-08-26-Real-Time-Maps-with-Event-Hubs-and-SignalR.md [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+date:   2015-08-26 16:23:28
+categories: Azure EventHubs Leaflet SignalR Geo Maps
+color: "blue"
+#image: "{{ site.baseurl }}/images/2015-08-26-Real-Time-Maps-with-Event-Hubs-and-SignalR/route_animation.gif" #should be ~350px tall
+excerpt: Real-time Maps with Event Hubs and SignalR Using Leaflet
+---
+
+Here on the Partner Catalyst team at Microsoft, we work with partners from all over the world on everything from long-term projects to week-long hackfests. 
+Recently, we spent a week in Hamburg, Germany, working with [Ströer](http://www.stroeer.com/) to help them tackle some thorny data visualization and data-handling problems. 
+They had a large corpus of geographic data on the foot-traffic patterns of people within the city (entirely anonymized), and wanted to gain some insight as to where traffic might congregate and when, getting a rough sense of "page views" for their physical advertising emplacements (e.g. billboards).
```
Comment (sedouard) [inline _posts/2015-08-26-Real-Time-Maps-with-Event-Hubs-and-SignalR.md]: Possibly a title heading here along the lines of 'Problem'

File: _posts/2015-08-26-Real-Time-Maps-with-Event-Hubs-and-SignalR.md [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+These were grandiose notions, to be sure, but the first step was to visualize the data. 
+As with all geographic data, the only way to get a proper sense of what it looks like is to visualize it on a map, so we needed a way to get real-time data from mobile applications to our map visualization.
+We built a [Cordova](http://cordova.apache.org/)-based mobile app driving geo-data into [Event Hubs](http://azure.microsoft.com/en-us/services/event-hubs/) (for future data), and a C#-based CLI for driving their existing data into Event Hubs as a spoof "real-time" feed. 
+Now that we had our geo-data, we built an [EventProcessorHost](https://www.nuget.org/packages/Microsoft.Azure.ServiceBus.EventProcessorHost) to consume it from within our ASP.Net website, do any required pre-processing, and drive the resulting data through to listening clients via [SignalR](http://signalr.net/).
+Our main client was a single-page web application using [Leaflet.js](http://leafletjs.com/) to render the most recent paths, and per-person and aggregated heatmaps.
```
Comment (sedouard) [inline _posts/2015-08-26-Real-Time-Maps-with-Event-Hubs-and-SignalR.md]: live link?

File: _posts/2015-08-26-Real-Time-Maps-with-Event-Hubs-and-SignalR.md [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+
+The ASP.Net application was [released on GitHub](https://github.com/noodlefrenzy/asp-mappy), and a more detailed walkthrough can be found [on my blog](http://www.mikelanzetta.com/2015/08/real-time-mapping-with-signalr-and-event-hubs/).
+
+![Screenshot of ASP.Net application]({{ site.baseurl }}/images/2015-08-26-Real-Time-Maps-with-Event-Hubs-and-SignalR/route_animation.gif)
+
+Future work involves integrating our more precise and stable Geodesy library for location binning and accurate heatmaps, and follow-on work involves driving a Spark pipeline from the EventHubs data to do additional aggregation and processing. 
+These works in progress will result in future case studies, but the [Geodesy library](https://github.com/juergenpf/Geodesy) is [out now on Nuget](http://www.nuget.org/packages/Geodesy/).
```
Comment (sedouard) [inline _posts/2015-08-26-Real-Time-Maps-with-Event-Hubs-and-SignalR.md]: code example of how to use this library would be fantastic

**Comment by sedouard** [comment]
In general I like code snippets and/or shiny things. Your blog post is really nice, if you bring some of that in (without necessarily repeating it) it would make it nice. Not a big deal though just thought I'd point it out.  I'll hold off on the magic words since our bender-bot will merge.

**Comment by cicorias** [comment]
Few edits but LGTM

**Comment by felixrieseberg** [comment]
@noodlefrenzy: Everything looks great here - let me know when you incorporated the changes so that I can merge

**Comment by noodlefrenzy** [comment]
Will do, @sedouard and @cicorias had some great, but time-consuming, feedback so it'll be a bit :)

File: _posts/2015-08-26-Real-Time-Maps-with-Event-Hubs-and-SignalR.md (hunk 2 above)
Comment (noodlefrenzy) [inline _posts/2015-08-26-Real-Time-Maps-with-Event-Hubs-and-SignalR.md]: All image links are, by definition, broken given the weird formatting we're supposed to do (according to the README). I'm assuming jekyll or something fixes things, but I'll let @felixrieseberg clue me in if I'm wrong.

**Comment by noodlefrenzy** [comment]
@cicorias and @sedouard - I've added section headers, an image w.r.t. data flow, some details/code on usage, and section headers. I didn't address @cicorias comment re: elevator pitch because I was under the impression we weren't doing those anymore but I'll seek clarification on that. I didn't address @sedouard comment re: live link because this isn't deployed anywhere at the moment (and when it was, it was showing private Stroeer data).

Please take another look and then I'll give Felix permission to push the button.

File: _posts/2015-08-26-Real-Time-Maps-with-Event-Hubs-and-SignalR.md [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+These works in progress will result in future case studies, but the [Geodesy library](https://github.com/juergenpf/Geodesy) is [out now on Nuget](http://www.nuget.org/packages/Geodesy/).  
+
+### Using the Website
+
+To use this application yourself, you just need to set up an appropriate Event Hub and drive traffic through it. 
+With the CLI I have checked in, it sends random point data through the pipeline and then renders it, so you don't even need a data source to get started.
+
+I've created a utility method that allows you to pull configuration data from either the cloud config, `web/app.config` or environment variables of the same name:
+
+```
```
Comment (sedouard) [inline _posts/2015-08-26-Real-Time-Maps-with-Event-Hubs-and-SignalR.md]: use ```cs to do c# syntax highlighting here.

File: _posts/2015-08-26-Real-Time-Maps-with-Event-Hubs-and-SignalR.md [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+        return _ConfigurationEntries.GetOrAdd(name, 
+            x => CloudConfigurationManager.GetSetting(x) ?? Environment.GetEnvironmentVariable(name));
+    }
+```
+
+This allows you to set environment variables with the configuration entries for your Event Hub, and then easily run the CLI and website without having to muck around in your app or web configs (or worry about checking in keys, in my case).
+
+To run both, you can see from `EventHubRoutePointSource` that you just need to set the following variables:
+
+```
```
Comment (sedouard) [inline _posts/2015-08-26-Real-Time-Maps-with-Event-Hubs-and-SignalR.md]: use ```cs to do c# syntax highlighting here.

**Comment by sedouard** [comment]
Besides adding the syntax highlighting LGTM

**Comment by noodlefrenzy** [comment]
@felixrieseberg I believe this is good to go.

**Comment by felixrieseberg** [comment]
@noodlefrenzy: Could you do me a quick favor [squash the commits](https://github.com/CatalystCode/case-studies/blob/gh-pages/git-tutorial.md#squashing-commits)?

**Pull request merged**